lib/fastly_nsq/manager.rb

```diff
@@ -1,104 +1,80 @@
 # frozen_string_literal: true
 
-require 'set'
-
 class FastlyNsq::Manager
-  attr_reader :listeners
+  DEADLINE = 30
+  DEFAULT_POOL_SIZE = 5
+
+  attr_reader :done, :pool, :logger
 
-  def initialize(options = {})
-    @options = options
-    @done = false
-    @listeners = Set.new
-    @plock = Mutex.new
+  def initialize(logger: FastlyNsq.logger, **pool_options)
+    @done      = false
+    @logger    = logger
+    @pool      = FastlyNsq::PriorityThreadPool.new(
+      { fallback_policy: :caller_runs, max_threads: DEFAULT_POOL_SIZE }.merge(pool_options),
+    )
   end
 
-  def start
-    setup_configured_listeners
-    @listeners.each(&:start)
+  def topic_listeners
+    @topic_listeners ||= {}
   end
 
-  def quiet
-    return if @done
-    @done = true
+  def topics
+    topic_listeners.keys
+  end
 
-    FastlyNsq.logger.info { 'Terminating quiet listeners' }
-    @listeners.each(&:terminate)
+  def listeners
+    topic_listeners.values
   end
 
-  PAUSE_TIME = 0.5
+  def terminate(deadline = DEADLINE)
+    return if done
 
-  def stop(deadline)
-    quiet
+    stop_listeners
 
-    sleep PAUSE_TIME
-    return if @listeners.empty?
+    return if pool.shutdown?
 
-    FastlyNsq.logger.info { 'Pausing to allow workers to finish...' }
-    remaining = deadline - Time.now
-    while remaining > PAUSE_TIME
-      return if @listeners.empty?
-      sleep PAUSE_TIME
-      remaining = deadline - Time.now
-    end
-    return if @listeners.empty?
+    stop_processing(deadline)
 
-    hard_shutdown
+    @done = true
   end
 
   def stopped?
-    @done
+    done
   end
 
-  def listener_stopped(listener)
-    @plock.synchronize do
-      @listeners.delete listener
+  def add_listener(listener)
+    logger.info { "Listening to topic:'#{listener.topic}' on channel: '#{listener.channel}'" }
+
+    if topic_listeners[listener.topic]
+      logger.warn("topic: #{listener.topic} was added more than once")
```

Comment:
this message should be in a block... so it doesn't make the string if we not setup to log at `warning` level.

Reply:
Good call.  Needs a test oo.